File: plus-one/Solution66.java
// https://leetcode.com/problems/plus-one/

import java.util.*;

class Solution66 {

    // T: O(n), S: O(n)
    public int[] plusOne(int[] digits) {
        int n = digits.length;
        for (int i = n - 1; i >= 0; i--) {
            if (digits[i] < 9) {
                digits[i]++;
                return digits;
            }
            digits[i] = 0;
        }
        var result = new int[n + 1];
        result[0] = 1;
        return result;
    }

    public static void main(String[] args) {
        var sol = new Solution66();
        var digits1 = new int[] { 1, 2, 3 };
        var digits2 = new int[] { 4, 3, 2, 1 };
        var digits3 = new int[] { 9 };
        System.out.println(Arrays.toString(sol.plusOne(digits1))); // [1,2,4]
        System.out.println(Arrays.toString(sol.plusOne(digits2))); // [4,3,2,2]
        System.out.println(Arrays.toString(sol.plusOne(digits3))); // [1,0]
    }

}

File: plus-one/solution66.go
// https://leetcode.com/problems/plus-one/

package main

import "fmt"

// T: O(n), S: O(n)
func plusOne(digits []int) []int {
	n := len(digits)
	for i := n - 1; i >= 0; i-- {
		if digits[i] < 9 {
			digits[i]++
			return digits
		}
		digits[i] = 0
	}
	result := make([]int, n+1)
	result[0] = 1
	return result
}

func main() {
	digits1 := []int{1, 2, 3}
	digits2 := []int{4, 3, 2, 1}
	digits3 := []int{9}
	fmt.Println(plusOne(digits1)) // [1,2,4]
	fmt.Println(plusOne(digits2)) // [4,3,2,2]
	fmt.Println(plusOne(digits3)) // [1,0]
}
